---
title: Pile d'appels et exceptions
description: Pile d'appels et exceptions
hide_table_of_contents: true
---

# Pile d'appels et exceptions

## Contenu

Video √† faire:
- pile d'appels dans le d√©bogueur
- ce qui se passe quand on lance une exception
- dans la pile d'appels, on peut laisser passer ou attraper
- syntaxe en kotlin
- si personne n'attrape, l'application crash, et on voit une trace (trace) de pile (stack) d'appels commun√©ment appele√© 
**stacktrace** 
- dans une application, on devrait avoir une fonction qui attrape. Si ce n'est pas le cas, faire un gros try catch dans le main
- dans une librairie, ce n'est pas le cas, on peut lancer des exceptions au programmeur de l'appli pour qu'il r√©agisse au probl√®me

Recette:
1. identifier les bouts de codes qui pourraient lancer une exception
2. se demander qui devrait l'attraper
3. si on ne sait pas, attraper au point d'entr√©e (main) et afficher un message d'erreur "une erreur inconnue est survenue"

Sources:
- https://kotlinlang.org/docs/exceptions.html

```kotlin
try {
    // some code
} catch (e: SomeException) {
    // handler
} finally {
    // optional finally block
}
```

Cette semaine nous discuterons d'abord de comment g√©rer des exceptions lanc√©es par le code qu'on utilise. Plus pr√©cisement, nous parlerons des 2 choix principaux:

- Attraper une exception (catch d'un "try catch") : attraper l'exception et g√©rer la situation
- Laisser passer l'exception. On esp√®re alors qu'une des m√©thodes de la pile d'appels va attraper.

Nous discuterons du fait qu'il faut attraper l'exception quelque part si on veut √©viter un crash de l'application.

Finalement nous parlerons de la possibilit√© de d√©finir des classes d'exceptions (il suffit d'√©tendre la classe Exception).

<Row>

<Column>

:::warning Avant la s√©ance (2h)

Vous regarderez le projet **[demo exceptions](https://github.com/departement-info-cem/3N5-Prog3/tree/main/code/Demo_Exceptions)**

Vous pouvez chercher les TODO pour trouver les explications importantes.

Lorsqu'on vous demande de chercher des TODO, vous devez t√©l√©charger le **[repo du cours](https://github.com/departement-info-cem/3N5-Prog3/tree/main)** (`Code > Download ZIP`) et retrouver le code dans la hi√©rarchie (il se trouve sous le r√©pertoire `code`). 

Pour lister les TODO dans IntelliJ : `Menu hamburger en haut √† gauche > View > Tool Windows > TODO`.

:::

</Column>

<Column>

:::tip Vid√©os

<Video url="https://youtu.be/bJ4MXns-aqo" />

:::

</Column>

<Column>

:::info √Ä faire pendant la s√©ance

- Compl√©ter les exercices de la semaine
- Continuer le **[premier travail](../tp/tp1)**

:::

</Column>

</Row>

:::note Exercices

Exercices type:
- expliquer un stack trace et sugg√©rer l'√©tape suivante
- sans ordi pr√©voir le comportement
- avec ordi et d√©bogueur, r√©pondre √† une question creuse
- programmer un bout de code qui illustre la compr√©hension

### Exercice devine qui va attraper

Fournir un exemple de code 

### Exercice TrouveLaValeur

Vous devez √©crire une classe appel√©e Trouveru. Cette classe doit contenir une m√©thode statique appel√©e trouve qui prend en param√®tre

- un tableau de String (String[]): la tableau dans lequel on cherche
- un String: l'√©l√©ment qu'on cherche

La m√©thode doit avoir le comportement suivant:

- si l'√©l√©ment est dans le tableau, vous renvoyez l'index de la premi√®re cellule o√π il se trouve
- si l'√©l√©ment n'est pas dans le tableau, vous lancez une exception. Vous devez d√©finir la classe d'exception dans une classe et choisir un nom d'exception significatif.

Finalement vous devez d√©montrer le fonctionnement de votre code dans une classe App avec un appel normal (o√π on trouve l'√©l√©ment) et un appel exceptionnel (qui lance l'exception).

### Exercice Bon ‚úîÔ∏è ou Non ‚ùå

Voici des situations. Dites si le code est bon, ou pas, et pourquoi en 1 ligne

#### Exemple 1

```java
public void uneFonction(){
	try {
		//du code‚Ä¶
	} catch (Exception e){
		System.out.println("Erreur : " + e);
	} catch (ArrayOutOfBoundsException a){
		System.out.println("Erreur : " + a);
	}
}
```

#### Exemple 2

```java
public void calculerPoidsSupernova() {
	try {
		//calculs compliqu√©s...
	} catch (NumberFormatException e) {}
}
```

### Exercice Attrape üé£ ou Transmet üì° ?

Voici des situations. Dites si on devrait attraper (**catch**) ou transmettre (**throws**) et pourquoi

#### Exemple 1

```java
public static void main() {
	try {
		calcul();
		lireFichier();
	} catch (ArithmeticException a) {
		println("Probl√®me dans les valeurs du calcul");
	} catch (IllegalArgumentException i ) {
		println("Probl√®me dans les param√®tres du programme");		
	} catch (InterruptedException t){
		println("Probl√®me avec le thread en cours");	
	}
}

public static void lireFichier(){
	// Code qui lit un fichier...
    // highlight-next-line
	// Q : Catch ou Throws?
}
```

#### Exemple 2

```java
public static void main( String[] args ){
	List<Integer> myList = Arrays.asList(1, 2, 3, 4, 5);
	List<List<Integer> > lists = Lists.partition(myList, -1);
			
	// Ce code utilise la librairie Guava.
    // highlight-next-line
	// Q : Catch ou Throws?
}
```

#### Exemple 3

```java
public void uneFonction(){
	etape1();
	etape2();
	etape3();
	etape4();
	etape5();
}

public void etape3(){
	//Code qui g√©n√®re une exception
    // highlight-next-line
	// Q : Catch ou Throws?
}
```

### Exercice sans machine üë®‚Äçüé§üë®‚Äçüíº

Sans ex√©cuter, selon toi est-ce que les 2 m√©thodes suivantes donnent le m√™me r√©sultat ou non?

Cr√©e toi un projet pour valider ta r√©ponse si les m√©thodes sont pareilles ou non.

TODO Comment ils doivent en garder une trace dans leur repo √©tudiant

```kotlin
public static void lireDesFichiersForTry(){
    for (int i = 0 ; i < 10 ; i++ ){
        try {
            lireFichier(i); 
        } catch (FileNotFoundException e) {
            System.out.println("fichier n'existe pas");
        }
    }
}

public static void lireDesFichiersTryFor(){
    try {
        for (int i = 0 ; i < 10 ; i++ ){
            lireFichier(i); 
        }
    } catch (FileNotFoundException e) {
        System.out.println("fichier n'existe pas");
    }
}	
```

:::